HSEARCH-499 Allow to filter elements to be indexed when mass-indexing (#2609)
hibernate/hibernate-search: documentation/src/main/asciidoc/reference/mapper-orm-indexing-massindexer.asciidoc

<2> Start the mass indexing process for the selected types and return when it is over.
====

> You can also select a subset of target entities to be reindexed.

Reviewer: This section is already quite long... maybe move your additions to a dedicated section, just above "`MassIndexer` parameters"?

E.g.:

```
[[mapper-orm-indexing-massindexer-conditional]]
== Conditional reindexing

[...]
```

> .Partial reindexing

Reviewer: I'd argue that "partial" is a bit confusing, since the example above, "Reindexing selected types using a `MassIndexer`", can also be considered as "partial" reindexing.

Maybe rather "conditional" reindexing?

Author: We already use the phrase `Partial reindexing` for documenting the similar feature for the batch JSR-352.
I'm going to change it for both.

====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmMassIndexerIT.java[tags=reindexOnly]
----
<1> Get the `SearchSession`.
<2> Create a `MassIndexer` targeting every indexed entity type.

> <3> Reindex only the books having year less or equals to 2100.

Reviewer: ```suggestion
<3> Reindex only the books published before year 2100.
```

> <4> Reindex only the author having birthday less of a given local date.

Reviewer: ```suggestion
<4> Reindex only the authors borns prior to a given local date.
```

> <5> For this case the date is passed as query parameter.

Reviewer: ```suggestion
<5> In this example the date is passed as a query parameter.
```

as explained in <<mapper-orm-indexing-massindexer-tuning-threads>>.
====

[[mapper-orm-indexing-massindexer-conditional]]
== Conditional reindexing
You can also select a subset of target entities to be reindexed.

.Use of Conditional reindexing
====
[source, JAVA, indent=0]
----
include::{sourcedir}/org/hibernate/search/documentation/mapper/orm/indexing/HibernateOrmMassIndexerIT.java[tags=reindexOnly]
----
<1> Get the `SearchSession`.
<2> Create a `MassIndexer` targeting every indexed entity type.
<3> Reindex only the books published before year 2100.
<4> Reindex only the authors born prior to a given local date.
<5> In this example the date is passed as a query parameter.
<6> Start the mass indexing process and return when it is over.
====

[WARNING]
====
Even if the reindexing is applied on a subset of entities, by default *all entities* will be purged at the start.

> At the moment there is no way to filter the entities that will be purged.

Reviewer: I just noticed this could lead people to think entities are always purged; maybe explain that purge can be disabled?

```suggestion
The purge <<mapper-orm-indexing-massindexer-parameters,can be disabled completely>>, but when enabled there is no way to filter the entities that will be purged.
```
